plugins/redis/redis.cpp

```diff
@@ -0,0 +1,904 @@
+    void clear(ICodeContext * ctx);
+    unsigned __int64 dbSize(ICodeContext * ctx);
+    bool exists(ICodeContext * ctx, const char * key);
+
+protected :
+    void parseOptions(ICodeContext * ctx, const char * _options);
+    void connect(ICodeContext * ctx, unsigned __int64 _database, const char * password);
+    void selectDB(ICodeContext * ctx, unsigned __int64 _database);
+    void authenticate(ICodeContext * ctx, const char * password);
+    void resetContextErr();
+    void readReply(Reply * reply);
+    void readReplyAndAssert(Reply * reply, const char * msg);
+    void readReplyAndAssertWithKey(Reply * reply, const char * msg, const char * key);
+    void assertKey(const redisReply * reply, const char * key);
+    void assertOnError(const redisReply * reply, const char * _msg);
+    void assertOnCommandError(const redisReply * reply, const char * cmd);
+    void assertOnCommandErrorWithDatabase(const redisReply * reply, const char * cmd);
+    void assertOnCommandErrorWithKey(const redisReply * reply, const char * cmd, const char * key);
+    void assertConnection();
+    void updateTimeout(unsigned __int64 _timeout);
+    void * allocateAndCopy(const char * src, size_t size);
+    bool isSameConnection(ICodeContext * ctx, const char * password) const;
+
+    //-------------------------------LOCKING------------------------------------------------
+    void handleLockOnSet(ICodeContext * ctx, const char * key, const char * value, size_t size, unsigned expire);
+    void handleLockOnGet(ICodeContext * ctx, const char * key, MemoryAttr * retVal, const char * password);
+    void encodeChannel(StringBuffer & channel, const char * key) const;
+    bool lock(ICodeContext * ctx, const char * key, const char * channel);
+    //--------------------------------------------------------------------------------------
+
+protected :
+    redisContext * context;
+    StringAttr options;
+    StringAttr ip;
+    unsigned serverIpPortPasswordHash;
+    int port;
+    unsigned __int64 timeout;
+    unsigned __int64 database;
+};
+
+//The following class is here to ensure destruction of the cachedConnection within the main thread
+//as this is not handled by the thread hook mechanism.
+static class mainThreadCachedConnection
+{
+public :
+    mainThreadCachedConnection() { }
+    ~mainThreadCachedConnection()
+    {
+        if (cachedConnection)
+            cachedConnection->Release();
+    }
+} mainThread;
+
+static void releaseContext()
+{
+    if (cachedConnection)
+    {
+        cachedConnection->Release();
+        cachedConnection = NULL;
+    }
+    if (threadHookChain)
+    {
+        (*threadHookChain)();
+        threadHookChain = NULL;
+    }
+}
+Connection::Connection(ICodeContext * ctx, const char * _options, unsigned __int64 _database, const char * password, unsigned __int64 _timeout)
+  : database(0), timeout(_timeout), port(0)
+{
+    serverIpPortPasswordHash = hashc((const unsigned char*)password, strlen(password), 0);
```

Comment:
minor: candidate for moving into a function and sharing with the call in isSameConnection